pydicom/valuerep.py

```diff
@@ -43,6 +43,319 @@
+DT_REGEX = re.compile(_range_regex(
+    r"\d{4}((0[1-9]|1[0-2])(([0-2]\d|3[01])(([01]\d|2[0-3])"
+    r"([0-5]\d((60|[0-5]\d)(\.\d{1,6} ?)?)?)?)?)?)?([+-][01]\d\d\d)?")
+)
+TM_REGEX = re.compile(_range_regex(
+    r"([01]\d|2[0-3])([0-5]\d((60|[0-5]\d)(\.\d{1,6} ?)?)?)?"))
+UR_REGEX = re.compile(r"^[A-Za-z_\d:/?#\[\]@!$&'()*+,;=%\-.~]* *$")
+
+
+def validate_vr_length(vr: str, value: Union[str, bytes]) -> Tuple[bool, str]:
+    """Validate the value length for a given VR.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str or bytes
+        The value to validate.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    max_length = MAX_VALUE_LEN.get(vr, 0)
+    if max_length > 0:
+        value_length = len(value)
+        if value_length > max_length:
+            return False, (
+                f"The value length ({value_length}) exceeds the "
+                f"maximum length of {max_length} allowed for VR {vr}."
+                f"Please see "
+                f"<https://dicom.nema.org/medical/dicom/current/output/html/"
+                f"part05.html#table_6.2-1> for allowed values for each VR."
+            )
+    return True, ""
+
+
+def validate_regex(vr: str, value: Union[str, bytes],
+                   regex: Any) -> Tuple[bool, str]:
+    """Validate the value for a given VR for allowed characters
+    using a regular expression.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str
+        The value to validate.
+    regex : re.Pattern
+        The precompiled regex pattern used for matching.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    if not re.match(regex, value):
+        return False, (
+            f"Invalid value for VR {vr}: '{value!r}'. Please see "
+            " <https://dicom.nema.org/medical/dicom/current/output"
+            "/html/part05.html#table_6.2-1> for allowed values for each VR."
+        )
+    return True, ""
+
+
+def validate_length_and_regex(vr: str, value: Union[str, bytes],
+                              regex: Any) -> Tuple[bool, str]:
+    """Validate the value for a given VR both for maximum length and
+    for allowed characters using a regular expression.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str
+        The value to validate.
+    regex : re.Pattern
+        The precompiled regex pattern used for matching.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    is_valid_len, msg1 = validate_vr_length(vr, value)
+    is_valid_expr, msg2 = validate_regex(vr, value, regex)
+    return is_valid_len and is_valid_expr, " ".join([msg1, msg2]).strip()
```

Comment:
I think I've wrecked your error message here by suggesting the link to the VR table, might be better to have a custom one rather than combining them

Reply:
Ah, I forgot about that message concatenation, thanks!